src/iarc7_sensors/orientation_filter.py

```diff
@@ -0,0 +1,184 @@
+        # propogated forward from the last OrientationAnglesStamped message
+        self._queue = [([initial_msg.data, float('NaN'), float('NaN')],
+                        initial_msg)]
+
+        self._transform_broadcaster = tf2_ros.TransformBroadcaster()
+
+        self._last_published_stamp = rospy.Time(0)
+
+    def _publish_transform(self, r, p, y, time):
+        transform_msg = TransformStamped()
+
+        transform_msg.header.stamp = time
+        transform_msg.header.frame_id = 'level_quad'
+        transform_msg.child_frame_id = 'quad'
+
+        quaternion = tf.transformations.quaternion_from_euler(r, p, y)
+        transform_msg.transform.rotation.x = quaternion[0]
+        transform_msg.transform.rotation.y = quaternion[1]
+        transform_msg.transform.rotation.z = quaternion[2]
+        transform_msg.transform.rotation.w = quaternion[3]
+
+        self._transform_broadcaster.sendTransform(transform_msg)
+
+    def _get_last_index(self, klass, time=None):
+        '''
+        Return the index of the last message of type `klass` (or the
+        last one before `time`)
+
+        Returns `-1` if there are no messages of type `klass` in the
+        queue
+        '''
+        for i in xrange(len(self._queue)-1, -1, -1):
+            if (type(self._queue[i][1]) == klass
+                 and (time is None or self._queue[i][1].header.stamp < time)):
+                return i
+        return -1
+
+    def _callback(self, klass, msg):
+        with self._lock:
+            last_i = self._get_last_index(klass)
+            if msg.header.stamp <= self._queue[last_i][1].header.stamp:
+                # This message isn't newer than the last one of type klass
+                rospy.logwarn(
+                    ('Ignoring message of type {} with timestamp {} older than '
+                   + 'previous message with timestamp {}')
+                        .format(
+                            klass,
+                            msg.header.stamp,
+                            self._queue[last_i][1].header.stamp))
+                return
+
+            for i in xrange(len(self._queue) - 1, -1, -1):
+                if self._queue[i][1].header.stamp < msg.header.stamp:
+                    index = i+1
+                    break
+            else:
+                index = 0
+
+            data = [float('NaN')] * 3
+            try:
+                data[1] = -msg.data.pitch
+                data[2] = msg.data.roll
+            except AttributeError:
+                # If this message type doesn't have pitch and roll, skip it
+                pass
+
+            self._queue.insert(index, (data, msg))
+            self._reprocess(index)
+
+            if (not math.isnan(self._queue[-1][0][0])
+                and not math.isnan(self._queue[-1][0][1])
+                and not math.isnan(self._queue[-1][0][2])):
+
+                new_stamp = max(self._last_published_stamp + rospy.Duration(0, 1),
```

Comment:
This could use a comment to explain that this stamp is being set 1 nano second into the future.